glib/src/value.rs

```diff
@@ -56,31 +51,262 @@ use crate::gstring::GString;
+
+impl<T: StaticType> GetValueError for WrongValueTypeOrNoneError<T> {
+    fn check(value: &Value) -> Result<(), Self> {
+        WrongValueTypeError::<T>::check(value)?;
+
+        unsafe {
+            // Values are always zero-initialized so even if pointers are only 32 bits then the
+            // whole 64 bit value will be 0 for NULL pointers.
+            if value.0.data[0].v_uint64 == 0 {
+                return Err(WrongValueTypeOrNoneError::UnexpectedNone);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl<T: StaticType> fmt::Debug for WrongValueTypeOrNoneError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            WrongValueTypeOrNoneError::WrongValueType(err) => f
+                .debug_tuple("WrongValueTypeOrNoneError::WrongValueType")
+                .field(err)
+                .finish(),
+            WrongValueTypeOrNoneError::UnexpectedNone => f
+                .debug_tuple("WrongValueTypeOrNoneError::UnexpectedNone")
+                .finish(),
+        }
+    }
+}
+
+impl<T: StaticType> fmt::Display for WrongValueTypeOrNoneError<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            WrongValueTypeOrNoneError::WrongValueType(err) => err.fmt(f),
+            WrongValueTypeOrNoneError::UnexpectedNone => write!(f, "Unexpected None",),
+        }
+    }
+}
+
+impl<T: StaticType> error::Error for WrongValueTypeOrNoneError<T> {}
+
+impl<T: StaticType> From<WrongValueTypeError<T>> for WrongValueTypeOrNoneError<T> {
+    fn from(err: WrongValueTypeError<T>) -> Self {
+        WrongValueTypeOrNoneError::WrongValueType(err)
+    }
+}
+
+/// Trait to retrieve the contained value from a `Value`.
+///
+/// Usually this would not be used directly but from the [`get`](struct.Value.html#method.get)
+/// function on a [`Value`](struct.Value.html)
+pub unsafe trait FromValue<'a>: Sized {
+    /// Error type if a value can't be retrieved.
+    type Error: GetValueError;
+
+    /// Get the contained value from a `Value`.
+    ///
+    /// # Safety
+    /// `Self::Error::check()` must be called first and must not fail.
+    unsafe fn from_value(value: &'a Value) -> Self;
+}
+
+/// Blanket implementation for all optional types.
+unsafe impl<'a, T: FromValue<'a, Error = WrongValueTypeOrNoneError<T>> + StaticType> FromValue<'a>
+    for Option<T>
+{
+    type Error = WrongValueTypeError<T>;
+
+    unsafe fn from_value(value: &'a Value) -> Self {
+        if let Err(WrongValueTypeOrNoneError::UnexpectedNone) = T::Error::check(value) {
+            None
+        } else {
+            Some(T::from_value(value))
+        }
+    }
+}
+
+/// Trait to convert a value to a `Value`.
+pub trait ToValue {
+    /// Convert a value to a `Value`.
+    fn to_value(&self) -> Value;
+
+    /// Returns the type identifer of `self`.
+    ///
+    /// This is the type of the value to be returned by `to_value`.
+    fn to_value_type(&self) -> Type;
```

Comment:
Suggestion: `value_type` or `type_`, since this is not a conversion.
Also, shouldn't that be an associative function?

Reply:
> Suggestion: value_type or type_, since this is not a conversion.

I've read it as "give the type that to_value returns". But `value_type()` is probably less confusing :)

Will change that, thanks :+1: 

>  Also, shouldn't that be an associative function?

No, then you could as well use `Self::static_type()`. This exists solely for trait objects so that on a `&dyn ToValue` you can check what the resulting type would be. An associated function is not object-safe, so you couldn't use `ToValue` as trait object anymore.

Reply:
> This exists solely for trait objects so that on a &dyn ToValue you can check what the resulting type would be.

Oh, I see!